### ISSUE TITLE
神经网络算法

### ISSUE THREAD

**Issue by dongzhebin** [dongzhebin/NN#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

# 设置matplotlib支持中文显示
matplotlib.rcParams['font.family'] = 'Arial Unicode MS' # 或 'SimHei', 'Microsoft YaHei'
matplotlib.rcParams['axes.unicode_minus'] = False

# --- 0. 定义文件路径 ---
# 请确保这里的路径指向你存放数据文件的实际位置
train_path = r"/Users/dongzhebin/Downloads/实验二+房价预测/house price/数据/train.csv"
test_path = r"/Users/dongzhebin/Downloads/实验二+房价预测/house price/数据/test.csv"
submission_path = r"/Users/dongzhebin/Downloads/实验二+房价预测/house price/数据/sample_submission.csv"

# --- 2.3 数据加载与探索 ---
print("--- 2.3 数据加载与探索 ---")
# 读取数据
train_df = pd.read_csv(train_path)
test_df = pd.read_csv(test_path)

print("训练集 前5行数据:")
print(train_df.head())
print("\n测试集 前5行数据:")
print(test_df.head())

print("\n训练集 基本信息:")
# train_df.info()
print(f"\n训练集 数据形状 (行, 列): {train_df.shape}")
print(f"测试集 数据形状 (行, 列): {test_df.shape}")

print("\n训练集 数值特征统计描述:")
print(train_df.describe())

# 可视化目标变量 SalePrice 的分布
plt.figure(figsize=(8, 5))
sns.histplot(train_df['SalePrice'], kde=True)
plt.title('房屋售价 (SalePrice) 分布')
plt.xlabel('售价')
plt.ylabel('频数')
plt.grid(True)
plt.show()

# --- 2.4 数据预处理 ---
print("\n--- 2.4 数据预处理 ---")

# 1. 特征选择：仅选择数值型特征
train_numeric = train_df.select_dtypes(include=[np.number])
test_numeric = test_df.select_dtypes(include=[np.number])

# 分离特征和目标变量
X_all_raw = train_numeric.drop(["SalePrice", "Id"], axis=1)
y_all = train_numeric["SalePrice"]
X_test_raw = test_numeric.drop(["Id"], axis=1)

# 确保测试集和训练集具有相同的特征列
common_cols = list(X_all_raw.columns)
X_test_raw = X_test_raw[common_cols]

print(f"选择数值特征后，训练特征形状: {X_all_raw.shape}")
print(f"选择数值特征后，测试特征形状: {X_test_raw.shape}")

# 2. 缺失值处理：使用中位数填充
median_values = X_all_raw.median()
X_all_filled = X_all_raw.fillna(median_values)
X_test_filled = X_test_raw.fillna(median_values)

print(f"\n填充后训练集缺失值总数: {X_all_filled.isnull().sum().sum()}")
print(f"填充后测试集缺失值总数: {X_test_filled.isnull().sum().sum()}")

# 3. 数据划分：将填充后的数据划分为训练集和验证集
X_train, X_val, y_train, y_val = train_test_split(
    X_all_filled, y_all, test_size=0.2, random_state=42
)
print(f"\n划分后：训练集特征 {X_train.shape}, 验证集特征 {X_val.shape}")

# 4. 特征缩放 (标准化 - 对神经网络很重要)
print("\n开始特征缩放 (StandardScaler)...")
# 评估用scaler：仅在训练集上fit，然后transform训练集和验证集
scaler_eval = StandardScaler()
X_train_scaled = scaler_eval.fit_transform(X_train)
X_val_scaled = scaler_eval.transform(X_val) # 使用在训练集上fit的scaler
print("用于评估的训练集和验证集缩放完成。")

# (准备最终预测用scaler：在所有训练数据上fit)
# scaler_final = StandardScaler()
# X_all_scaled_final = scaler_final.fit_transform(X_all_filled)
# X_test_scaled_final = scaler_final.transform(X_test_filled)
# print("用于最终预测的全部训练数据和测试集缩放完成。")

# --- 2.5.2 神经网络模型 ---
print("\n--- 2.5.2 神经网络模型 ---")

# 1. 获取输入维度
input_dim = X_train_scaled.shape[1]
print(f"神经网络输入层维度: {input_dim}")

# 2. 构建模型结构 (Sequential API)
model_nn = keras.Sequential(
    [
        layers.Dense(128, activation='relu', input_shape=(input_dim,), name="INPUT_LAYER"),
        layers.Dropout(0.3, name="Dropout_1"), # Dropout 防止过拟合
        layers.Dense(64, activation='relu', name="HIDDEN_LAYER_1"),
        layers.Dropout(0.3, name="Dropout_2"),
        layers.Dense(32, activation='relu', name="HIDDEN_LAYER_2"),
        layers.Dense(1, name="OUTPUT_LAYER") # 回归任务输出层通常只有1个单元且无激活函数
    ],
    name="HousePricePredictionNN"
)

print("\n神经网络模型结构概要:")
model_nn.summary() # 打印模型结构

# 3. 编译模型：指定优化器、损失函数和评估指标
model_nn.compile(optimizer='adam',    # Adam 优化器
                   loss='mse',        # 均方误差损失函数 (Mean Squared Error)
                   metrics=['mae'])   # 同时监控平均绝对误差 (Mean Absolute Error)

# 4. 训练模型
epochs = 500      # 训练轮数
batch_size = 50  # 每批样本数
print(f"\n开始训练神经网络，共 {epochs} 轮...")

# validation_data 用于在每轮结束后在验证集上评估性能
history = model_nn.fit(X_train_scaled, y_train,
                       epochs=epochs,
                       batch_size=batch_size,
                       validation_data=(X_val_scaled, y_val),
                       verbose=1) # verbose=1 会显示训练进度条和每轮结果

print("神经网络训练完成。")

# 5. 在验证集上预测 (注意 predict 返回值可能需要展平)
y_val_pred_nn_raw = model_nn.predict(X_val_scaled)
y_val_pred_nn = y_val_pred_nn_raw.flatten() # 展平成一维数组

# 6. 评估模型
# 从 history 对象获取训练过程中的最终验证集损失 (MSE)
# final_val_loss = history.history['val_loss'][-1]
# print(f"神经网络 - 训练结束时验证集 Loss (MSE): {final_val_loss:.4f}")

# 重新计算验证集的 RMSE 和 R² 以便与其他模型比较
rmse_nn = np.sqrt(mean_squared_error(y_val, y_val_pred_nn))
r2_nn = r2_score(y_val, y_val_pred_nn)
print(f"\n神经网络 - 验证集 RMSE: {rmse_nn:.4f}")
print(f"神经网络 - 验证集 R2: {r2_nn:.4f}")

# 7. 可视化：训练过程中的损失变化
history_dict = history.history
train_loss = history_dict['loss']
val_loss = history_dict['val_loss']

plt.figure(figsize=(10, 5))
plt.plot(range(1, epochs + 1), train_loss, 'b-', label='训练损失 (MSE)')
plt.plot(range(1, epochs + 1), val_loss, 'r-', label='验证损失 (MSE)')
plt.title('神经网络训练过程 - 损失变化')
plt.xlabel('Epochs')
plt.ylabel('均方误差 (MSE)')
plt.legend()
plt.grid(True)
plt.show()

# 8. 可视化：真实 vs 预测散点图 (验证集)
plt.figure(figsize=(8, 6))
plt.scatter(y_val, y_val_pred_nn, alpha=0.6, color='green', label='预测值')
plt.plot([y_val.min(), y_val.max()], [y_val.min(), y_val.max()], 'r--', label='理想预测 (y=x)')
plt.xlabel("真实房价 (验证集)")
plt.ylabel("预测房价 (神经网络)")
plt.title("神经网络模型 - 真实值 vs 预测值")
plt.legend()
plt.grid(True)
plt.show()

# 9. 可视化：残差分布图 (验证集)
residuals_nn = y_val - y_val_pred_nn # 注意 y_val_pred_nn 已经展平
plt.figure(figsize=(8, 5))
sns.histplot(residuals_nn, bins=30, kde=True, color='green')
plt.title("神经网络模型 - 残差分布图")
plt.xlabel("残差 (真实值 - 预测值)")
plt.ylabel("频数")
plt.axvline(0, color='red', linestyle='--', linewidth=1)
plt.grid(True)
plt.show()

# --- 最终预测与提交 ---
print("\n--- 生成神经网络预测提交文件 ---")

# 准备最终预测数据：需要在 *全部* 训练数据上拟合 Scaler
print("为最终预测重新进行数据缩放...")
scaler_final = StandardScaler()
X_all_scaled_final = scaler_final.fit_transform(X_all_filled) # 在填充后的全部训练数据上拟合和转换
X_test_scaled_final = scaler_final.transform(X_test_filled) # 使用同一个 scaler 转换测试集
print("最终预测数据缩放完成。")

# 重新构建并训练最终模型 (使用所有训练数据)
# 注意：理论上应该用相同结构和超参数，也可以选择直接用上面训练好的 model_nn
# 但重新训练可以确保模型在所有数据上学习
print("重新构建并训练最终神经网络模型...")
input_dim_final = X_all_scaled_final.shape[1]
model_nn_final = keras.Sequential(
    [
        layers.Dense(128, activation='relu', input_shape=(input_dim_final,), name="INPUT_LAYER"),
        layers.Dropout(0.3, name="Dropout_1"),
        layers.Dense(64, activation='relu', name="HIDDEN_LAYER_1"),
        layers.Dropout(0.3, name="Dropout_2"),
        layers.Dense(32, activation='relu', name="HIDDEN_LAYER_2"),
        layers.Dense(1, name="OUTPUT_LAYER")
    ],
    name="HousePricePredictionNN-final"
)
model_nn_final.compile(optimizer='adam', loss='mse', metrics=['mae'])
# 不设置 validation_data，因为是在所有数据上训练
model_nn_final.fit(X_all_scaled_final, y_all, epochs=epochs, batch_size=batch_size, verbose=1)
print("最终模型训练完成。")

# 对处理后的测试集进行预测
print("在测试集上进行最终预测...")
predictions_nn_raw = model_nn_final.predict(X_test_scaled_final)
predictions_nn = predictions_nn_raw.flatten() # 展平

# 生成提交文件
submission_nn = pd.DataFrame({
    "Id": test_df["Id"], # 使用原始测试集的Id
    "SalePrice": predictions_nn
})
submission_nn.to_csv("C:/Users/PeterCao/Desktop/DONG0428/submission_nn.csv", index=False)
print("神经网络预测结果已保存为 submission_nn.csv")

print("\n神经网络流程结束。")
